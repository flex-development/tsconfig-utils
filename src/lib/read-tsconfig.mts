/**
 * @file readTsconfig
 * @module tsconfig-utils/lib/readTsconfig
 */

import withTrailingSlash from '#internal/with-trailing-slash'
import {
  cwd as cwdUrl,
  getSource,
  moduleResolve,
  toUrl,
  type ModuleId
} from '@flex-development/mlly'
import pathe from '@flex-development/pathe'
import type { Tsconfig } from '@flex-development/tsconfig-types'
import type {
  ReadTsconfigOptions,
  ResolvedTsconfig
} from '@flex-development/tsconfig-utils'
import { isObjectCurly } from '@flex-development/tutils'
import JSON from 'json5'
import stripBom from 'strip-bom'

/**
 * Resolve and read a tsconfig file.
 *
 * @see {@linkcode ModuleId}
 * @see {@linkcode ReadTsconfigOptions}
 * @see {@linkcode Tsconfig}
 *
 * @async
 *
 * @this {void}
 *
 * @param {ModuleId} id
 *  Module id of tsconfig file, or path to tsconfig file
 * @param {ReadTsconfigOptions | null | undefined} [options]
 *  Read options
 * @return {Promise<ResolvedTsconfig | null>}
 *  Resolved tsconfig or `null` if tsconfig file is not found
 */
async function readTsconfig(
  this: void,
  id: ModuleId,
  options?: ReadTsconfigOptions | null | undefined
): Promise<ResolvedTsconfig | null> {
  /**
   * URL of current working directory.
   *
   * @const {URL} cwd
   */
  const cwd: URL = toUrl(withTrailingSlash(options?.cwd ?? cwdUrl()))

  /**
   * Module extensions to probe for.
   *
   * @const {Set<string>} extensions
   */
  const extensions: Set<string> = new Set<string>(['.json', '.jsonc', '.json5'])

  /**
   * URL of parent module.
   *
   * @const {URL} parent
   */
  const parent: URL = new URL('parent.json', cwd)

  /**
   * Module specifiers to try resolving.
   *
   * @const {[string, ModuleId][]} tries
   */
  const tries: [string, ModuleId][] = [
    [String(id), options?.parent ?? parent],
    [pathe.dot + pathe.sep + String(id), parent]
  ]

  /**
   * URL of tsconfig file.
   *
   * @var {URL | null} url
   */
  let url: URL | null = null

  for (const [specifier, parent] of tries) {
    try {
      url = await moduleResolve(
        specifier,
        parent,
        options?.conditions,
        options?.mainFields,
        options?.preserveSymlinks,
        options?.fs
      )
    } catch {
      url = null

      if (!extensions.has(pathe.extname(specifier))) {
        for (const ext of extensions) {
          try {
            url = await moduleResolve(
              pathe.addExt(specifier, ext),
              parent,
              options?.conditions,
              options?.mainFields,
              options?.preserveSymlinks,
              options?.fs
            )

            break
          } catch {
            url = null
          }
        }
      }
    }

    if (url !== null && url.protocol !== 'node:') {
      /**
       * Resolved tsconfig.
       *
       * @const {ResolvedTsconfig} resolved
       */
      const resolved: ResolvedTsconfig = { tsconfig: {}, url }

      /**
       * Contents of tsconfig file.
       *
       * @const {string | null | undefined} contents
       */
      const contents: string | null | undefined = await getSource(url, options)

      if (typeof contents === 'string') {
        resolved.tsconfig = JSON.parse(stripBom(contents.trim() || 'null'))
        if (!isObjectCurly(resolved.tsconfig)) resolved.tsconfig = {}
      }

      return resolved
    }
  }

  return null
}

export default readTsconfig
