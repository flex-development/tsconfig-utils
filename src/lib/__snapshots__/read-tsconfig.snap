// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/readTsconfig > should return resolved tsconfig object ("#fixtures/tsconfig.json") 1`] = `
{
  "tsconfig": {
    "compilerOptions": {
      "allowImportingTsExtensions": true,
      "allowJs": true,
      "alwaysStrict": true,
      "customConditions": [
        "tsconfig-utils",
        "development",
      ],
      "declaration": true,
      "declarationMap": true,
      "lib": [
        "es2023",
      ],
      "module": "esnext",
      "moduleResolution": "bundler",
      "newLine": "lf",
      "noEmit": true,
      "noErrorTruncation": true,
      "noUncheckedSideEffectImports": true,
      "noUnusedParameters": false,
      "paths": {
        "https://esm.sh/lodash@4.17.21": [
          "node_modules/@types/lodash/index.d.ts",
        ],
      },
      "resolveJsonModule": true,
      "sourceMap": true,
      "target": "esnext",
      "verbatimModuleSyntax": true,
    },
    "exclude": [
      "coverage",
      "dist",
      "node_modules",
    ],
    "extends": "@tsconfig/strictest/tsconfig.json",
    "include": [
      "**/**.json",
      "**/**.mjs",
      "**/**.mts",
      "**/**.ts",
    ],
    "mdx": {
      "plugins": [
        "remark-directive",
        "remark-frontmatter",
        "remark-gfm",
      ],
    },
  },
  "url": file://\${process.cwd()}/__fixtures__/tsconfig.json,
}
`;

exports[`unit:lib/readTsconfig > should return resolved tsconfig object ("@tsconfig/strictest/tsconfig.json") 1`] = `
{
  "tsconfig": {
    "$schema": "https://json.schemastore.org/tsconfig",
    "_version": "2.0.0",
    "compilerOptions": {
      "allowUnreachableCode": false,
      "allowUnusedLabels": false,
      "checkJs": true,
      "esModuleInterop": true,
      "exactOptionalPropertyTypes": true,
      "isolatedModules": true,
      "noFallthroughCasesInSwitch": true,
      "noImplicitOverride": true,
      "noImplicitReturns": true,
      "noPropertyAccessFromIndexSignature": true,
      "noUncheckedIndexedAccess": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "skipLibCheck": true,
      "strict": true,
    },
  },
  "url": file://\${process.cwd()}/node_modules/@tsconfig/strictest/tsconfig.json,
}
`;

exports[`unit:lib/readTsconfig > should return resolved tsconfig object ("__fixtures__/tsconfig.empty") 1`] = `
{
  "tsconfig": {},
  "url": file://\${process.cwd()}/__fixtures__/tsconfig.empty.json,
}
`;
